proposals/0000-local-modules.rst

```diff
@@ -194,143 +195,294 @@ Motivation
+
+  In a module import/export list, a symbolic name that begins with a ``:`` is
+  in the type constructor namespace; otherwise, it is in the term variable
+  namespace. (This is an unusual juxtaposition.)
+
+* **Qualified name**: A qualified name is a name prepended with a module name
+  followed by a ``.``, with no intervening whitespace. Both alphanumeric and
+  symbolic names can be qualified.
+  
+* **Entity**: An entity is a definition that can be exported and imported.
+  An entity has a name. Entities include variables, classes, datatypes, and
+  constructors, among a few other constructs.
+
+* **Environment**: An environment is a set of names and qualified names,
+  used for validating name occurrences.
+
+* **in scope**: A name is said to be **in scope** when it is in the current
+  environment.
+  
+* **Scope**: A scope is a lexical region of a program where all name occurrences are
+  looked up in an environment (or a superset of an environment).
+
+* **Module**: A module is a set of declarations of entities, along with a set
+  of exported names.
+  
+* **Exports**: The exports from a module is a set of names. Exports are always *unqualified* names.
+  (These names can be written with a qualification
+  to disambiguate, if the unqualified name has multiple entries in the environment
+  of the exporting module. But the exported name itself is unqualified.)
+
+* **Imports**: An ``import`` statement names a module and (optionally) a set
+  of *unqualified* names. (This set may be specified via complement, using ``hiding``;
+  this proposal will not worry about this detail.)
+  These names are brought into scope in some fashion:
+
+  * If the import uses ``as``, the module name appearing after the ``as``
+    is known as the module alias. If there is no ``as``, the module name itself
+    is used as the alias.
+
+  * The set of names indicated is brought into scope (that is, put into the
+    global environment), qualified by the module alias in the import.
+
+  * If the import does not use ``qualified``, the names are additionally
+    brought into scope unqualified.
+
+Core Proposal Change Specification
+----------------------------------
+
+This proposal's specification is divided into pieces. This core piece is a necessary
+component of the overall proposal. Later pieces can be chosen piecemeal.
+
+1. A qualified name can now have any number of module qualifications, each
+   separated by a ``.`` and with no intervening whitespace. This can lead
+   to potential ambiguity: is the name ``Data.List.length`` qualified by
+   ``Data.List`` (a well-formed module name) or qualified both by ``Data``
+   and by ``List``? If ``length`` is in scope qualified both by ``Data.List``
+   and by both ``Data`` and ``List``, the occurrence ``Data.List.length`` is
+   an error. There is no way to disambiguate locally without using module
+   aliases or other renamings. (Programmers should seek to avoid this scenario.)
```

Comment:
Is this equivalent to making the two cases (long module name vs several qualifications) distinguishable?

Reply:
It depends on what you mean by "distinguishable". A long module name (e.g. `Data.Set`) is distinct from a composite module name (like what you would get with `module Data where module Set where fromList = 5`). But this proposal offers no way for users to disambiguate between these. That is, if you have the two examples above mixed together and write `Data.Set.fromList`, you will get an ambiguity error, which has no resolution other than renaming something.